Add a-ball handwired trackball based on ADSN9800

brickbots opened the pull request:
It's a trackball running QMK, with no buttons at all!  Designed to be used as an adjunct to another keyboard to provide pointing capabilities, while the keyboard itself provides mouse button functionality.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

brickbots commented:
> This should be refactored to use https://docs.qmk.fm/#/spi_driver

Thank you @zvecr for pointing out the QMK driver here.  I've refactored the code to use it 👍

zvecr commented on keyboards/handwired/aball/adns.h:
```diff
@@ -0,0 +1,16 @@
+
```
```suggestion
#pragma once

```

zvecr commented on keyboards/handwired/aball/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,41 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+#include "print.h"
+#include "adns.h"
+
```
```suggestion

```

zvecr commented on keyboards/handwired/aball/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,41 @@
+
+  [_QWERTY] = LAYOUT( KC_C,   KC_Q )
+};
+
+
+
+void keyboard_post_init_user(void) {
+  // Customise these values to desired behaviour
+  //debug_enable=true;
+  //debug_matrix=true;
+  //debug_keyboard=true;
+  //debug_mouse=true;
+  //print("KB INIT COMPLETE");
+  //setPinOutput(F5);
+  return;
+}
+
```
```suggestion

```

zvecr commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,267 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "spi_master.h"
+#include "spi_master.c"
```
```suggestion
```

zvecr commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,36 @@
+
+# Add trackball support
+POINTING_DEVICE_ENABLE = yes
+SRC += adns.c
```
```suggestion
SRC += adns.c spi_master.c
```

brickbots commented:
Thank you @zvecr for your patient guidance and suggestions!  I've actioned all of your suggestions and pushed this branch again. Cheers!

fauxpark commented on keyboards/handwired/aball/adns9800_srom_A6.h:
```diff
@@ -0,0 +1,3078 @@
+#pragma once
+
+#include <avr/pgmspace.h>
```
```suggestion
#include "progmem.h"
```

fauxpark commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,266 @@
+}
+
+void pointing_device_init(void) {
+    print("STSARTING INTI\n");
```
```suggestion
    dprint("STARTING INTI\n");
```

fauxpark commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,266 @@
+ */
+#include "spi_master.h"
+#include "adns.h"
+#include "print.h"
```
```suggestion
#include "debug.h"
```

fauxpark commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,266 @@
+    adns_write(REG_Configuration_I, 0x08);
+
+    wait_ms(100);
+    print("INIT ENDED\n");
```
```suggestion
    dprint("INIT ENDED\n");
```

brickbots commented:
Hi @fauxpark . Thank you for your suggestions to use the newer debug features rather than print.  Much appreciated! :+1:

tzarc commented:
My only reservation with merging this is the firmware -- there's no license or comments as to where it came from.

brickbots commented:
Thanks @tzarc for the review.  I've added the best information I could find about this, it seems to have been sent to the original author, MrJohnK, directly by Avago as byte code that he formatted as a c byte array for arduino use.   I hope this additional info is sufficient to help people track it back to the source.

Also, apologies for the sloppy merge there, I switched computers and had to remerge a few of the previous print->debug suggestions.  Sorry for the mess!

drashna commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,36 @@
+
+# Add trackball support
+POINTING_DEVICE_ENABLE = yes
+SRC += adns.c spi_master.c
```
Since it may be sensitive to timing, I would rather see SPI master added like this: 

```suggestion
SRC += adns.c 
QUANTUM_LIB_SRC =+ spi_master.c
```

This should prevent any issues if LTO is enabled.

brickbots commented:
Thank you @drashna ! I'll have to dig into the functional change here to really understand it, but I very much appreciate your suggestion about the import directives!

brickbots commented:
Hello Reviewers!  Sorry if I have lost the plot here a bit.  I don't see any pending changes, but I could have missed something along the way.  Is there something holding up this merge request that I might be able to address?

drashna commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,263 @@
+    delta_y = 0;
+
+    pointing_device_set_report(report);
+    pointing_device_send();
```
Just a heads up, without any check here about changes in the report, this will keep the system awake.

This is an issue I ran into with the pimironi trackball module.  
If you're polling, it will cause this, but if you're using interrupts, it shouldn't be an issue.

drashna replied:
Also, the other issue here is that there is no way to customize the behavior (such as holding down a switch to switch to a "scrolling" mode, for instance)

brickbots replied:
Thank you @drashna!   The heads up about keeping the machine awake makes sense as it keeps getting mouse reports. I've wrapped this so that it only occurs on change.

As for the customization... This is a bit of a strange beast, it's a trackball with NO buttons.. just a ball.  I think you have a Kyria, so you may appreciate my inspiration here:
https://github.com/brickbots/aball

drashna replied:
Welcome! Personally I just check if any change has occurred in the report, and not send if no change has occurred. Either way works, I think.

And that's nice! (And yup, I have a kyria)

brickbots commented:
Hi reviewers!

Thank you for all the feedback and guidance so far.  I'm keen to get this merged if possible, or closed if not.  Is the semi-mysterious ADSN firmware a deal-breaker here?

drashna commented:
the ADSN firmware thing should be fine.

brickbots commented:
Hi @fauxpark and @zvecr !

I appreciate both of your time and help so far.  If I have missed a comment or change request somewhere along the line I apologize, but I'm not sure what needs to be done to get this PR merged.  Would it be please be possible to get more notes, or an approval, on this pull request?

Thanks!

drashna commented on keyboards/handwired/aball/adns.c:
```diff
@@ -0,0 +1,262 @@
+/* Copyright 2020 Richard Sutherland <rich@brickbots.com>
```
This can be removed.

drashna commented on keyboards/handwired/aball/adns.h:
```diff
@@ -0,0 +1,32 @@
+/* Copyright 2020 Richard Sutherland <rich@brickbots.com>
```
Also, can be removed.

drashna commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,37 @@
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
```suggestion
```

drashna commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,37 @@
+EXTRAKEY_ENABLE = no       # Audio control and System control
+CONSOLE_ENABLE = no        # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
+
```
```suggestion
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output

```

drashna commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,37 @@
+
+# Add trackball support
+POINTING_DEVICE_ENABLE = yes
+SRC += adns.c
+QUANTUM_LIB_SRC += spi_master.c
```
```suggestion
POINTING_DEVICE_DRIVER = adns9800
```

brickbots commented:
Thanks @drashna for the suggestions to bring this PR up to snuff with the recent develop branch changes.  It's exciting to see first class support for the ADNS9800 sensor!

drashna commented:
> Thanks @drashna for the suggestions to bring this PR up to snuff with the recent develop branch changes. It's exciting to see first class support for the ADNS9800 sensor!

Welcome!  And honestly, I'm happy to add first class support for as many sensors as I can. 

Also, it may be worth adding some of the defines from: 
https://github.com/qmk/qmk_firmware/pull/15304

I've been using those code changes, and they're fantastic.

fauxpark commented on keyboards/handwired/aball/rules.mk:
```diff
@@ -0,0 +1,22 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = no       # Audio control and System control
+CONSOLE_ENABLE = no        # Console for debug
+COMMAND_ENABLE = no        # Commands for debug and configuration
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output on port C6
```
```suggestion
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

fauxpark commented on keyboards/handwired/aball/config.h:
```diff
@@ -0,0 +1,54 @@
+#define DEVICE_VER 0x0001
+#define MANUFACTURER Richard Sutherland
+#define PRODUCT ABall
+#define DESCRIPTION Buttonless trackball designed for use with a keyboard
```
```suggestion
```

fauxpark reviewed (changes requested):
Missing info.json.

brickbots commented:
Thanks @fauxpark for the feedback on the feature description changes and the reminder about info.json!  I appreciate it 👍

fauxpark commented on keyboards/handwired/aball/info.json:
```diff
@@ -0,0 +1,10 @@
+        "vid": "0xFEED",
+        "pid": "0xC2F0",
+        "device_ver": "0x0001"
+    }
```
Missing layouts.

brickbots replied:
Yes...indeed.... but it has no buttons :-)  I can put a null there to indicate so?

drashna replied:
Not really, but take a look at the ploopyco/trackball_nano to see how I handled the lack of any keys.

fauxpark replied:
You have a `LAYOUT` macro with two buttons in aball.h.

brickbots replied:
Aha, yes, thank you @fauxpark!  When I set this up I was not sure how to handle a matrix-less device and get it to compile.  I've now emulated what was done with the Ploopy nano and I hope it's a lot clearer and will pass the linter now as well.  Thanks for your patience!

noroadsleft reviewed (approved):
Looks good to me.

noroadsleft commented:
Thanks!
